src/transform/src/canonicalization/flat_map_elimination.rs

```diff
@@ -42,70 +45,49 @@ impl crate::Transform for FlatMapElimination {
+                    Some(Ok(datum)) => args.push(datum),
+                    // Give up if any arg is not a literal, or if it's a literal error.
+                    _ => return,
                 }
-            } else if is_supported_unnest(func) {
-                let func = func.clone();
-                let exprs = exprs.clone();
-                use mz_expr::MirScalarExpr;
-                use mz_repr::RowArena;
-                if let MirScalarExpr::Literal(Ok(row), ..) = &exprs[0] {
-                    let temp_storage = RowArena::default();
-                    if let Ok(mut iter) = func.eval(&[row.iter().next().unwrap()], &temp_storage) {
-                        match (iter.next(), iter.next()) {
-                            (None, _) => {
-                                // If there are no elements in the literal argument, no output.
-                                relation.take_safely(None);
-                            }
-                            (Some((row, Diff::ONE)), None) => {
-                                assert_eq!(func.output_type().column_types.len(), 1);
-                                *relation =
-                                    input.take_dangerous().map(vec![MirScalarExpr::Literal(
-                                        Ok(row),
-                                        func.output_type().column_types[0].clone(),
-                                    )]);
-                                if with_ordinality {
-                                    *relation =
-                                        relation.take_dangerous().map_one(MirScalarExpr::literal(
-                                            Ok(Datum::Int64(1)),
-                                            ScalarType::Int64,
-                                        ));
-                                }
-                            }
-                            _ => {}
-                        }
-                    };
+            }
+            let temp_storage = RowArena::new();
+            let (first, second) = match func.eval(&args, &temp_storage) {
+                Ok(mut r) => (r.next(), r.next()),
+                // don't play with errors
+                Err(_) => return,
+            };
+            match (first, second) {
+                // The table function evaluated to an empty collection.
+                (None, None) => {
+                    relation.take_safely(None);
                 }
+                // The table function evaluated to a collection with exactly 1 row.
+                (Some((first_row, Diff::ONE)), None) => {
+                    let types = func.output_type().column_types;
+                    let map_exprs = first_row
+                        .into_iter()
+                        .zip_eq(types)
+                        .map(|(d, typ)| MirScalarExpr::Literal(Ok(Row::pack_slice(&[d])), typ))
+                        .collect();
+                    *relation = input.take_dangerous().map(map_exprs);
```

Comment:
Maybe premature defensive coding, but we might want a `return;` after setting the relation? The method should not continue once we swap the relation out, and .. tbh probably Rust will defend against that, but just like with the `Wrap` case we don't want to let execution continue after this point.

Reply:
Added a `return`.

Reply:
Yeah, clippy alternates between not understanding software engineering and having helpful thoughts about it.